File: core/src/main/java/com/egoal/darkestpixeldungeon/items/scrolls/Scroll.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015  Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2016 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.egoal.darkestpixeldungeon.items.scrolls;

import android.util.Pair;

import com.egoal.darkestpixeldungeon.actors.buffs.Pressure;
import com.egoal.darkestpixeldungeon.actors.hero.Hero;
import com.egoal.darkestpixeldungeon.items.artifacts.UnstableSpellbook;
import com.egoal.darkestpixeldungeon.Badges;
import com.egoal.darkestpixeldungeon.actors.buffs.Blindness;
import com.egoal.darkestpixeldungeon.items.Item;
import com.egoal.darkestpixeldungeon.items.ItemStatusHandler;
import com.egoal.darkestpixeldungeon.messages.Messages;
import com.egoal.darkestpixeldungeon.sprites.HeroSprite;
import com.egoal.darkestpixeldungeon.sprites.ItemSpriteSheet;
import com.egoal.darkestpixeldungeon.utils.GLog;
import com.watabou.utils.Bundle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

public abstract class Scroll extends Item {

  public static final String AC_READ = "READ";

  protected static final float TIME_TO_READ = 1f;

  protected int initials;

  private static final Class<?>[] scrolls = {
          ScrollOfIdentify.class,
          ScrollOfMagicMapping.class,
          ScrollOfRecharging.class,
          ScrollOfRemoveCurse.class,
          ScrollOfTeleportation.class,
          ScrollOfUpgrade.class,
          ScrollOfRage.class,
          ScrollOfTerror.class,
          ScrollOfLullaby.class,
          ScrollOfEnchanting.class,
          ScrollOfPsionicBlast.class,
          ScrollOfMirrorImage.class,
          ScrollOfCurse.class,
          ScrollOfLight.class,
  };

  private static final HashMap<String, Integer> runes = new HashMap<String,
          Integer>() {
    {
      put("KAUNAN", ItemSpriteSheet.SCROLL_KAUNAN);
      put("SOWILO", ItemSpriteSheet.SCROLL_SOWILO);
      put("LAGUZ", ItemSpriteSheet.SCROLL_LAGUZ);
      put("YNGVI", ItemSpriteSheet.SCROLL_YNGVI);
      put("GYFU", ItemSpriteSheet.SCROLL_GYFU);
      put("RAIDO", ItemSpriteSheet.SCROLL_RAIDO);
      put("ISAZ", ItemSpriteSheet.SCROLL_ISAZ);
      put("MANNAZ", ItemSpriteSheet.SCROLL_MANNAZ);
      put("NAUDIZ", ItemSpriteSheet.SCROLL_NAUDIZ);
      put("BERKANAN", ItemSpriteSheet.SCROLL_BERKANAN);
      put("ODAL", ItemSpriteSheet.SCROLL_ODAL);
      put("TIWAZ", ItemSpriteSheet.SCROLL_TIWAZ);
      put("QI", ItemSpriteSheet.SCROLL_QI);
      put("LINGEL", ItemSpriteSheet.SCROLL_LINGEL);
    }
  };

  private static ItemStatusHandler<Scroll> handler;

  private String rune;

  public boolean ownedByBook = false;

  {
    stackable = true;
    defaultAction = AC_READ;

    cursed = false;
    cursedKnown = true;
  }

  @SuppressWarnings("unchecked")
  public static void initLabels() {
    handler = new ItemStatusHandler<>((Class<? extends Scroll>[]) scrolls,
            runes);
  }

  public static void save(Bundle bundle) {
    handler.save(bundle);
  }

  public static void saveSelectively(Bundle bundle, ArrayList<Item> items) {
    handler.saveSelectively(bundle, items);
  }

  @SuppressWarnings("unchecked")
  public static void restore(Bundle bundle) {
    handler = new ItemStatusHandler<>((Class<? extends Scroll>[]) scrolls,
            runes, bundle);
  }

  public Scroll() {
    super();
    reset();
  }

  @Override
  public void reset() {
    super.reset();
    image = handler.image(this);
    rune = handler.label(this);
  }

  ;

  @Override
  public ArrayList<String> actions(Hero hero) {
    ArrayList<String> actions = super.actions(hero);
    actions.add(AC_READ);
    return actions;
  }

  @Override
  public void execute(Hero hero, String action) {

    super.execute(hero, action);

    if (action.equals(AC_READ)) {

      if (hero.buff(Blindness.class) != null) {
        GLog.w(Messages.get(this, "blinded"));
      } else if (hero.buff(UnstableSpellbook.bookRecharge.class) != null
              && hero.buff(UnstableSpellbook.bookRecharge.class).isCursed()
              && !(this instanceof ScrollOfRemoveCurse)) {
        GLog.n(Messages.get(this, "cursed"));
      } else {
        // want read
        kotlin.Pair<Boolean, String> pr = hero.canRead();
        if (!pr.getFirst()) GLog.n(pr.getSecond());
        else {
          curUser = hero;
          curItem = detach(hero.getBelongings().backpack);
          doRead();
        }
      }

    }
  }

  abstract protected void doRead();

  protected void readAnimation() {
    curUser.spend(TIME_TO_READ);
    curUser.busy();
    ((HeroSprite) curUser.sprite).read();
  }

  public boolean isKnown() {
    return handler.isKnown(this);
  }

  public void setKnown() {
    if (!isKnown() && !ownedByBook) {
      handler.know(this);
    }

    Badges.INSTANCE.validateAllScrollsIdentified();
  }

  @Override
  public Item identify() {
    setKnown();
    return super.identify();
  }

  @Override
  public String name() {
    return isKnown() ? name : Messages.get(Scroll.class, rune);
  }

  @Override
  public String info() {
    return isKnown() ?
            desc() :
            Messages.get(this, "unknown_desc");
  }

  public Integer initials() {
    return isKnown() ? initials : null;
  }

  @Override
  public boolean isUpgradable() {
    return false;
  }

  @Override
  public boolean isIdentified() {
    return isKnown();
  }

  public static HashSet<Class<? extends Scroll>> getKnown() {
    return handler.known();
  }

  public static HashSet<Class<? extends Scroll>> getUnknown() {
    return handler.unknown();
  }

  public static boolean allKnown() {
    return handler.known().size() == scrolls.length;
  }

  @Override
  public int price() {
    return 30 * quantity;
  }
}

File: core/src/main/java/com/egoal/darkestpixeldungeon/items/scrolls/Scroll.kt
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015  Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2016 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.egoal.darkestpixeldungeon.items.scrolls

import com.egoal.darkestpixeldungeon.actors.hero.Hero
import com.egoal.darkestpixeldungeon.items.artifacts.UnstableSpellbook
import com.egoal.darkestpixeldungeon.Badges
import com.egoal.darkestpixeldungeon.Challenge
import com.egoal.darkestpixeldungeon.actors.buffs.Blindness
import com.egoal.darkestpixeldungeon.items.Item
import com.egoal.darkestpixeldungeon.items.ItemStatusHandler
import com.egoal.darkestpixeldungeon.messages.M
import com.egoal.darkestpixeldungeon.messages.Messages
import com.egoal.darkestpixeldungeon.sprites.HeroSprite
import com.egoal.darkestpixeldungeon.sprites.ItemSpriteSheet
import com.egoal.darkestpixeldungeon.utils.GLog
import com.watabou.utils.Bundle

import java.util.ArrayList
import java.util.HashSet

abstract class Scroll : Item() {
    protected var initials: Int = 0

    private var rune: String = ""

    var ownedByBook = false

    val isKnown: Boolean
        get() = handler.isKnown(this)

    override val isUpgradable: Boolean
        get() = false

    override val isIdentified: Boolean
        get() = isKnown

    init {
        stackable = true
        defaultAction = AC_READ

        cursed = false
        cursedKnown = true
    }

    init {
        reset()
    }

    override fun reset() {
        super.reset()
        image = handler.image(this)
        rune = handler.label(this)
    }

    override fun actions(hero: Hero): ArrayList<String> {
        val actions = super.actions(hero)
        actions.add(AC_READ)
        return actions
    }

    override fun execute(hero: Hero, action: String) {
        super.execute(hero, action)

        if (action == AC_READ) {
            if (hero.buff(Blindness::class.java) != null) {
                GLog.w(Messages.get(this, "blinded"))
            } else if (hero.buff(UnstableSpellbook.bookRecharge::class.java)?.isCursed == true &&
                    this !is ScrollOfRemoveCurse) {
                GLog.n(Messages.get(this, "cursed"))
            } else {
                // want read
                val (first, second) = hero.canRead()
                if (!first)
                    GLog.n(second)
                else {
                    curUser = hero
                    curItem = detach(hero.belongings.backpack)!!
                    doRead()
                }
            }

        }
    }

    override fun doPickUp(hero: Hero): Boolean {
        if (hero.challenges.contains(Challenge.PathOfAsceticism)) {
            GLog.n(M.L(Challenge::class.java, "gone", name()))
            hero.next()
            return true
        }

        return super.doPickUp(hero)
    }

    protected abstract fun doRead()

    protected fun readAnimation() {
        curUser.spend(TIME_TO_READ)
        curUser.busy()
        (curUser.sprite as HeroSprite).read()
    }

    fun setKnown() {
        if (!isKnown && !ownedByBook) {
            handler.know(this)
        }

        Badges.validateAllScrollsIdentified()
    }

    override fun identify(): Item {
        setKnown()
        return super.identify()
    }

    override fun name(): String {
        return if (isKnown) name else M.L(Scroll::class.java, rune)
    }

    override fun info(): String = if (isKnown) desc() else M.L(this, "unknown_desc")

    fun initials(): Int? = if (isKnown) initials else null

    override fun price(): Int = 30 * quantity

    companion object {
        const val AC_READ = "READ"

        const val TIME_TO_READ = 1f

        private val scrolls = arrayOf(
                ScrollOfIdentify::class.java, ScrollOfMagicMapping::class.java, ScrollOfRecharging::class.java,
                ScrollOfRemoveCurse::class.java, ScrollOfTeleportation::class.java, ScrollOfUpgrade::class.java,
                ScrollOfRage::class.java, ScrollOfTerror::class.java, ScrollOfLullaby::class.java,
                // ScrollOfEnchanting::class.java,
                ScrollOfPsionicBlast::class.java, ScrollOfMirrorImage::class.java,
                ScrollOfCurse::class.java, ScrollOfLight::class.java)

        private val runes = hashMapOf<String, Int>(
                "KAUNAN" to ItemSpriteSheet.SCROLL_KAUNAN,
                "SOWILO" to ItemSpriteSheet.SCROLL_SOWILO,
                "LAGUZ" to ItemSpriteSheet.SCROLL_LAGUZ,
                "YNGVI" to ItemSpriteSheet.SCROLL_YNGVI,
                "GYFU" to ItemSpriteSheet.SCROLL_GYFU,
                "RAIDO" to ItemSpriteSheet.SCROLL_RAIDO,
                "ISAZ" to ItemSpriteSheet.SCROLL_ISAZ,
                "MANNAZ" to ItemSpriteSheet.SCROLL_MANNAZ,
                "NAUDIZ" to ItemSpriteSheet.SCROLL_NAUDIZ,
                "BERKANAN" to ItemSpriteSheet.SCROLL_BERKANAN,
                "ODAL" to ItemSpriteSheet.SCROLL_ODAL,
                "TIWAZ" to ItemSpriteSheet.SCROLL_TIWAZ,
                // "QI" to ItemSpriteSheet.SCROLL_QI,
                "LINGEL" to ItemSpriteSheet.SCROLL_LINGEL
        )

        private lateinit var handler: ItemStatusHandler<Scroll>

        fun initLabels() {
            handler = ItemStatusHandler(scrolls, runes)
        }

        fun save(bundle: Bundle) {
            handler.save(bundle)
        }

        fun saveSelectively(bundle: Bundle, items: ArrayList<Scroll>) {
            handler.saveSelectively(bundle, items)
        }

        fun restore(bundle: Bundle) {
            handler = ItemStatusHandler(scrolls, runes, bundle)
        }

        val known: HashSet<Class<out Scroll>>
            get() = handler.known()

        val unknown: HashSet<Class<out Scroll>>
            get() = handler.unknown()

        fun allKnown(): Boolean = handler.known().size == scrolls.size
    }
}
